Start work on a MuseScore plugin.

cghawthorne opened the pull request:
Currently has the ability to reconstruct a selected region as a midi
file.

cghawthorne committed:
New score proto

cghawthorne committed:
Start work on a MuseScore plugin.

Currently has the ability to reconstruct a selected region as a midi
file.

cghawthorne committed:
Updated comments and reverted score-relative times

cghawthorne committed:
fix name

cghawthorne committed:
change segment start time name

cghawthorne committed:
Merge branch 'master' into scoreproto

czhuang committed:
Merge pull request #1 from cghawthorne/scoreproto

New score proto

czhuang commented on magenta/musescore/magenta.qml:
```diff
@@ -0,0 +1,109 @@
+          keySignatures: [],
+          tempos: [],
+          notes: [],
+          totalTime: curScore.duration,
```
Do we want to store both total time of score and selected duration?

cghawthorne replied:
I'll add a wrapper proto to include the user-selected portion of the score that the model should fill in.

czhuang commented on magenta/musescore/magenta.qml:
```diff
@@ -0,0 +1,109 @@
+          tempos: [],
+          notes: [],
+          totalTime: curScore.duration,
+          sourceType: 'SCORE',
```
Do we want to know it's from MuseScore?

cghawthorne replied:
I'll modify the source field in the proto to include a string name in addition to the source enum.

czhuang commented on magenta/musescore/magenta.qml:
```diff
@@ -0,0 +1,109 @@
+           } else {
+             endTick = cursor.tick;
+           }
+           endStaff = cursor.staffIdx;
```
What happens if there's multiple staves selected?

czhuang replied:
What happens when multiple staves are selected?  Does cursor.rewind(2) go to the bottom of the staff or has a default staff it would go to?  I only remember in time it would rewind to the end of piece (I don't have a mac with me to check it out MuseScore, and might be confusing between rewind(1) or rewind(2).

czhuang commented on magenta/musescore/magenta.qml:
```diff
@@ -0,0 +1,109 @@
+                      elem.globalDuration.ticks / 480 * cursor.tempo;
+
+                  var nsNote = {};
+                  nsNote.pitch = scoreNote.ppitch;
```
this is midi pitch of the note?  add tonal pitch class later? https://musescore.org/en/plugin-development/tonal-pitch-class-enum
and to NoteSequence proto?

cghawthorne replied:
Yeah, I plan to do that in a followup PR.

cghawthorne committed:
Hard code tempos and key signatures, add part info

Access to tempos and key signatures from MuseScore plugins is currently
broken. If we need to specify it, we can later add a dialog for users to
input tempo and key signature information. Or we can hope it gets fixed
soon.

Here are the relevant issues:
https://musescore.org/en/node/115971
https://musescore.org/en/node/100501

cghawthorne committed:
Improve score element extraction, add fill params.

Use some workarounds to extract time signatures and key signature
changes from MuseScore. Still haven't found a way to get tempo changes
until 2.0.4 is released.

Switched to using an AutoFillRequest message which includes the entire
score plus parameters describing what section should be filled by the
model.

Included additional source information.
